lib/switchtec.c

```diff
@@ -60,90 +60,100 @@
-	{0x4084, SWITCHTEC_GEN4, SWITCHTEC_PFX},   //PFX 84XG4
-	{0x4068, SWITCHTEC_GEN4, SWITCHTEC_PFX},   //PFX 68XG4
-	{0x4052, SWITCHTEC_GEN4, SWITCHTEC_PFX},   //PFX 52XG4
-	{0x4036, SWITCHTEC_GEN4, SWITCHTEC_PFX},   //PFX 36XG4
-	{0x4028, SWITCHTEC_GEN4, SWITCHTEC_PFX},   //PFX 28XG4
-	{0x4100, SWITCHTEC_GEN4, SWITCHTEC_PSX},   //PSX 100XG4
-	{0x4184, SWITCHTEC_GEN4, SWITCHTEC_PSX},   //PSX 84XG4
-	{0x4168, SWITCHTEC_GEN4, SWITCHTEC_PSX},   //PSX 68XG4
-	{0x4152, SWITCHTEC_GEN4, SWITCHTEC_PSX},   //PSX 52XG4
-	{0x4136, SWITCHTEC_GEN4, SWITCHTEC_PSX},   //PSX 36XG4
-	{0x4128, SWITCHTEC_GEN4, SWITCHTEC_PSX},   //PSX 28XG4
-	{0x4200, SWITCHTEC_GEN4, SWITCHTEC_PAX},   //PAX 100XG4
-	{0x4284, SWITCHTEC_GEN4, SWITCHTEC_PAX},   //PAX 84XG4
-	{0x4268, SWITCHTEC_GEN4, SWITCHTEC_PAX},   //PAX 68XG4
-	{0x4252, SWITCHTEC_GEN4, SWITCHTEC_PAX},   //PAX 52XG4
-	{0x4236, SWITCHTEC_GEN4, SWITCHTEC_PAX},   //PAX 36XG4
-	{0x4228, SWITCHTEC_GEN4, SWITCHTEC_PAX},   //PAX 28XG4
+	{0x8531, SWITCHTEC_PFX},   //PFX 24xG3
+	{0x8532, SWITCHTEC_PFX},   //PFX 32xG3
+	{0x8533, SWITCHTEC_PFX},   //PFX 48xG3
+	{0x8534, SWITCHTEC_PFX},   //PFX 64xG3
+	{0x8535, SWITCHTEC_PFX},   //PFX 80xG3
+	{0x8536, SWITCHTEC_PFX},   //PFX 96xG3
+	{0x8541, SWITCHTEC_PSX},   //PSX 24xG3
+	{0x8542, SWITCHTEC_PSX},   //PSX 32xG3
+	{0x8543, SWITCHTEC_PSX},   //PSX 48xG3
+	{0x8544, SWITCHTEC_PSX},   //PSX 64xG3
+	{0x8545, SWITCHTEC_PSX},   //PSX 80xG3
+	{0x8546, SWITCHTEC_PSX},   //PSX 96xG3
+	{0x8551, SWITCHTEC_PAX},   //PAX 24XG3
+	{0x8552, SWITCHTEC_PAX},   //PAX 32XG3
+	{0x8553, SWITCHTEC_PAX},   //PAX 48XG3
+	{0x8554, SWITCHTEC_PAX},   //PAX 64XG3
+	{0x8555, SWITCHTEC_PAX},   //PAX 80XG3
+	{0x8556, SWITCHTEC_PAX},   //PAX 96XG3
+	{0x8561, SWITCHTEC_PFXL},  //PFXL 24XG3
+	{0x8562, SWITCHTEC_PFXL},  //PFXL 32XG3
+	{0x8563, SWITCHTEC_PFXL},  //PFXL 48XG3
+	{0x8564, SWITCHTEC_PFXL},  //PFXL 64XG3
+	{0x8565, SWITCHTEC_PFXL},  //PFXL 80XG3
+	{0x8566, SWITCHTEC_PFXL},  //PFXL 96XG3
+	{0x8571, SWITCHTEC_PFXI},  //PFXI 24XG3
+	{0x8572, SWITCHTEC_PFXI},  //PFXI 32XG3
+	{0x8573, SWITCHTEC_PFXI},  //PFXI 48XG3
+	{0x8574, SWITCHTEC_PFXI},  //PFXI 64XG3
+	{0x8575, SWITCHTEC_PFXI},  //PFXI 80XG3
+	{0x8576, SWITCHTEC_PFXI},  //PFXI 96XG3
+	{0x4000, SWITCHTEC_PFX},   //PFX 100XG4
+	{0x4084, SWITCHTEC_PFX},   //PFX 84XG4
+	{0x4068, SWITCHTEC_PFX},   //PFX 68XG4
+	{0x4052, SWITCHTEC_PFX},   //PFX 52XG4
+	{0x4036, SWITCHTEC_PFX},   //PFX 36XG4
+	{0x4028, SWITCHTEC_PFX},   //PFX 28XG4
+	{0x4100, SWITCHTEC_PSX},   //PSX 100XG4
+	{0x4184, SWITCHTEC_PSX},   //PSX 84XG4
+	{0x4168, SWITCHTEC_PSX},   //PSX 68XG4
+	{0x4152, SWITCHTEC_PSX},   //PSX 52XG4
+	{0x4136, SWITCHTEC_PSX},   //PSX 36XG4
+	{0x4128, SWITCHTEC_PSX},   //PSX 28XG4
+	{0x4200, SWITCHTEC_PAX},   //PAX 100XG4
+	{0x4284, SWITCHTEC_PAX},   //PAX 84XG4
+	{0x4268, SWITCHTEC_PAX},   //PAX 68XG4
+	{0x4252, SWITCHTEC_PAX},   //PAX 52XG4
+	{0x4236, SWITCHTEC_PAX},   //PAX 36XG4
+	{0x4228, SWITCHTEC_PAX},   //PAX 28XG4
```

Comment:
I don't see the point of this. I prefer using the device ID table unless we can't actually get the device ID; then fall back on generation. But then we'd also need to do something similar for the variant....

Reply:
Previously our device only has one boot phase -- main firmware. Now we introduce the new boot phase BL1 and BL2. At any time, the device boot phase can be one of (BL1), BL2, or main firmware. 

Unfortunately, some of the information previously available in main firmware are not available in BL1/2 phase (device not fully initialized yet):
- The PING command is available in all phases, and can be used reliably to get the device generation
- Device ID is only available in main firmware phase
- Variant is determined from device ID, and can only be determined in main firmware as well

Reply:
Ok, well lets only fall back on the ping command if we can't get a sensible device ID ...

Reply:
If device is not in main firmware phase, function 'get_device_id()' would fail miserably. We have to avoid calling 'get_device_id()' in this case. That's why we want to get current boot phase first. 

I agree this function name is not accurate, something like get_info might be a better fit.

Reply:
I don't see how it will fail "miserably"... it returns a value from a register. Worst case the value is not correct, and it's easy to detect that sucha value is not in our table.

If the firmware is crashing or anything like that due to reading that register that is a huge problem that needs to be fixed in the firmware.

Reply:
I see what you mean. I will change the code to get device ID first, and fall back to ping if it fails.